# 📊 Stack Overflow Developer Survey Analysis

## Project Overview
This project analyzes the Stack Overflow Developer Survey dataset using Python and Pandas.

The goal is to extract meaningful insights from a large, real-world dataset.

## Skills Demonstrated
- Data cleaning
- Data exploration
- Data filtering
- GroupBy analysis
- Statistical analysis
- Real-world data interpretation

In [46]:
import pandas as pd

In [47]:
df = pd.read_csv("survey_results_public.csv", low_memory=False)

In [48]:
df.shape

(49191, 172)

In [49]:
df.head()

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,...,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,Vertex AI,NaN,NaN,NaN,ChatGPT,NaN,When I don’t trust AI’s answers,"Troubleshooting, profiling, debugging",61256.0,10.0
1,2,I am a developer by profession,25-34 years old,"Associate degree (A.A., A.S., etc.)",Employed,NaN,2.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers;When I want to...,All skills. AI is a flop.,104413.0,9.0
2,3,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Independent contractor, freelancer, or self-em...",None of the above,10.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code;GitHub Copilot;Google Gemini,NaN,When I don’t trust AI’s answers;When I want to...,"Understand how things actually work, problem s...",53061.0,8.0
3,4,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,None of the above,4.0,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code,NaN,When I don’t trust AI’s answers;When I want to...,NaN,36197.0,6.0
4,5,I am a developer by profession,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)","Independent contractor, freelancer, or self-em...","Caring for dependents (children, elderly, etc.)",21.0,"No, I am not new to coding and did not learn n...",NaN,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers,"critical thinking, the skill to define the tas...",60000.0,7.0


In [50]:
df.columns

Index(['ResponseId', 'MainBranch', 'Age', 'EdLevel', 'Employment',
       'EmploymentAddl', 'WorkExp', 'LearnCodeChoose', 'LearnCode',
       'LearnCodeAI',
       ...
       'AIAgentOrchestration', 'AIAgentOrchWrite', 'AIAgentObserveSecure',
       'AIAgentObsWrite', 'AIAgentExternal', 'AIAgentExtWrite', 'AIHuman',
       'AIOpen', 'ConvertedCompYearly', 'JobSat'],
      dtype='object', length=172)

## Question 1: How many people participated in the survey?

### Approach
Each row represents one respondent, so counting rows gives total participants.

In [51]:
df.shape[0]

49191

### Insight
A total of 49,191 respondents participated in the survey. This is a large dataset, which improves the reliability of insights.

*________________________________________________________________________________________________________________________________*

## Question 2: How many respondents answered all questions?

### Approach
- We need to count the number of rows that do not contain any missing values (NaN). 
- These rows represent respondents who answered all questions.

In [52]:
df.dropna().shape[0]

0

In [53]:
important_cols = ['Age', 'WorkExp', 'Employment', 'EdLevel']

df[important_cols].dropna().shape[0]

42818

### Insight

No respondents answered all 172 questions in the dataset, indicating a high level of missing data, which is common in large-scale surveys.

Relying on complete-case analysis would result in losing all data. Therefore, it is more effective to focus on a subset of important columns to retain a meaningful number of observations for analysis.

*________________________________________________________________________________________________________________________________*

## Question 3: What are the central tendency measures of WorkExp?

### Approach

We analyze the WorkExp (years of experience) column to calculate central tendency measures.

Since the column may contain missing or non-numeric values, we first convert it to numeric format and remove NaN values.

### Then, we calculate:
- Mean (average experience)
- Median (middle value)
- Mode (most frequent value)

In [54]:
df['WorkExp']

0         8.0
1         2.0
2        10.0
3         4.0
4        21.0
         ... 
49186     NaN
49187    25.0
49188    17.0
49189     2.0
49190     NaN
Name: WorkExp, Length: 49191, dtype: float64

In [55]:
df['WorkExp'].dtype

dtype('float64')

In [56]:
df['WorkExp'] = pd.to_numeric(df['WorkExp'], errors='coerce')

In [57]:
workexp_clean = df['WorkExp'].dropna()

In [58]:
mean_exp = workexp_clean.mean()
median_exp = workexp_clean.median()
mode_exp = workexp_clean.mode()

In [59]:
mean_exp, median_exp, mode_exp

(np.float64(13.367402606485907),
 np.float64(10.0),
 0    10.0
 Name: WorkExp, dtype: float64)

### Insight

The average work experience is approximately 13.4 years, while both the median and mode are 10 years.

This indicates that the distribution is right-skewed, meaning there are some respondents with very high experience levels that increase the average.

Most respondents have around 10 years of experience, indicating a strong concentration of mid-level developers in the dataset.

*________________________________________________________________________________________________________________________________*

## Question 4: How many people work remotely?

### Approach
We analyze the RemoteWork column to identify respondents who work remotely.

Since the column contains text values, we use string matching to filter rows that include "Remote" or similar categories (e.g., Hybrid).

We then count the number of matching respondents.

In [60]:
df.columns

Index(['ResponseId', 'MainBranch', 'Age', 'EdLevel', 'Employment',
       'EmploymentAddl', 'WorkExp', 'LearnCodeChoose', 'LearnCode',
       'LearnCodeAI',
       ...
       'AIAgentOrchestration', 'AIAgentOrchWrite', 'AIAgentObserveSecure',
       'AIAgentObsWrite', 'AIAgentExternal', 'AIAgentExtWrite', 'AIHuman',
       'AIOpen', 'ConvertedCompYearly', 'JobSat'],
      dtype='object', length=172)

In [61]:
df['RemoteWork'].value_counts()

RemoteWork
Remote                                                                          10931
Hybrid (some remote, leans heavy to in-person)                                   6732
In-person                                                                        6042
Hybrid (some in-person, leans heavy to flexibility)                              5831
Your choice (very flexible, you can come in when you want or just as needed)     4244
Name: count, dtype: int64

In [62]:
# Extended definition of remote work including flexible arrangements #

# Fully remote only
remote_count = df[df['RemoteWork'] == 'Remote'].shape[0]

# Remote + Hybrid
remote_hybrid_count = df[df['RemoteWork'].str.contains('Remote|Hybrid', na=False)].shape[0]

# Remote + Hybrid + Flexible
remote_flexible_count = df[df['RemoteWork'].str.contains('Remote|Hybrid|flexible', case=False, na=False)].shape[0]

In [63]:
remote_count

10931

In [64]:
remote_hybrid_count

23494

In [65]:
remote_flexible_count

27738

### Insight

A substantial number of respondents work fully remotely (10,931 participants).

When including hybrid work arrangements, the total rises to 23,494 respondents, indicating that flexible work models are widely adopted in the tech industry.

Expanding the definition further to include flexible work options increases the count to 27,738 respondents, highlighting the strong shift toward adaptable work environments.

This suggests that hybrid and flexible work models are more common than strictly fully remote arrangements.

*________________________________________________________________________________________________________________________________*

## Question 5: What percentage of respondents use Python?

### Approach

The programming languages column contains multiple values separated by semicolons (;).

To identify Python users, we check whether the string "Python" appears in each row.

We then count how many respondents used Python and calculate the percentage relative to the total number of participants.


In [66]:
python_users = df[df['LanguageHaveWorkedWith'].str.contains('Python', na=False)]

In [67]:
python_users_count = python_users.shape[0]
python_users_count

18466

In [68]:
total_count = df.shape[0]
total_count

49191

In [69]:
python_percentage = (python_users_count / total_count) * 100
python_percentage

37.53938728629221

In [70]:
round(python_percentage, 2)

37.54

### Insight

Approximately **37.54%** of respondents reported using Python.

This indicates that more than one-third of developers in the survey use Python, highlighting its strong adoption across the industry.

Python’s popularity is likely driven by its versatility in data analysis, automation, machine learning, and backend development.

*________________________________________________________________________________________________________________________________*

## Question 6: How many people learned programming through online courses?

### Approach

The learning methods column contains multiple values separated by semicolons (;).

To identify respondents who learned programming through online courses, we check whether the string "Online" appears in each row.

We then count the number of matching respondents.

In [71]:
online_learners = df[df['LearnCode'].str.contains('Online', na=False)]

In [72]:
online_learners_count = online_learners.shape[0]
online_learners_count

10973

In [73]:
online_percentage = (online_learners_count / total_count) * 100
round(online_percentage, 2)

22.31

### Insight

Approximately **22.31%** of respondents reported learning programming through online courses.

This suggests that while online education is a significant learning path, the majority of developers still rely on alternative methods such as formal education, self-study, or on-the-job experience.

The result highlights the growing but not yet dominant role of online learning platforms in developer education.

*________________________________________________________________________________________________________________________________*

## Question 7: What are the average and median yearly salaries of Python developers across countries?

### Approach

To analyze salary differences among Python users across countries:

- First, we filter respondents who reported using Python.
- Then, we select relevant columns: Country and yearly compensation.
- We convert the salary column to numeric and remove missing values.
- To ensure reliable results, we exclude countries with very small sample sizes.
- Finally, we group the data by country and calculate:
  - Mean (average salary)
  - Median (middle value)

This approach allows us to compare salary distributions across different countries while avoiding misleading results caused by insufficient data.

In [74]:
python_users = df[df['LanguageHaveWorkedWith'].str.contains('Python', na=False)]

In [75]:
python_users = python_users[['Country', 'ConvertedCompYearly']]

In [76]:
python_users['ConvertedCompYearly'] = pd.to_numeric(
    python_users['ConvertedCompYearly'],
    errors='coerce'
)

In [77]:
python_users_clean = python_users.dropna()

In [78]:
country_salary = python_users_clean.groupby('Country')['ConvertedCompYearly'].agg(['mean', 'median'])

In [79]:
country_salary_sorted = country_salary.sort_values(by='mean', ascending=False)
country_salary_sorted.head(10)

,mean,median
Country,,
Oman,390135.000000,390135.0
Andorra,226103.500000,226103.5
Viet Nam,218837.166667,8254.0
United States of America,173298.590211,150000.0
Switzerland,156456.600000,142592.0
Singapore,147515.121951,91391.0
Israel,135828.365385,142594.0
Nigeria,128773.276596,1808.0
Ireland,120523.918919,116015.0


In [80]:
country_counts = python_users_clean['Country'].value_counts()

In [81]:
valid_countries = country_counts[country_counts > 50].index

In [82]:
python_users_filtered = python_users_clean[
    python_users_clean['Country'].isin(valid_countries)
]

In [83]:
country_salary = python_users_filtered.groupby('Country')['ConvertedCompYearly'].agg(['mean', 'median'])

country_salary_sorted = country_salary.sort_values(by='mean', ascending=False)

country_salary_sorted.head(10)

,mean,median
Country,,
United States of America,173298.590211,150000.0
Switzerland,156456.600000,142592.0
Israel,135828.365385,142594.0
Ireland,120523.918919,116015.0
Australia,118091.410423,97514.0
United Kingdom of Great Britain and Northern Ireland,117662.362667,94958.5
Denmark,115672.425000,98289.0
Norway,104690.125000,89439.0
Canada,102846.026423,87550.0


### Insight

After filtering out countries with small sample sizes, we observe that the highest salaries among Python developers are concentrated in developed countries.

The **United States** leads with the highest average salary (~$173K), followed by Switzerland and Israel, indicating strong compensation in tech-driven economies.

European countries such as Ireland, the United Kingdom, Denmark, and Norway also show consistently high median salaries, suggesting stable and well-distributed income levels.

The differences between mean and median values indicate some level of income inequality, but overall trends remain consistent.

This analysis highlights that geographic location plays a significant role in developer compensation, even within the same technology stack.

*________________________________________________________________________________________________________________________________*

## Question 8: What are the education levels of the top 5 highest-paid respondents?

### Approach

To identify the education levels of the highest-paid respondents:

- We select relevant columns: education level, country, and yearly compensation.
- Convert the salary column to numeric format and remove missing values.
- Sort the dataset in descending order based on yearly compensation.
- Select the top 5 highest-paid respondents.

This allows us to examine the educational background of top earners in the dataset.

In [90]:
top_earners = df[['EdLevel', 'Country', 'ConvertedCompYearly']].copy()

In [91]:
top_earners['ConvertedCompYearly'] = pd.to_numeric(
    top_earners['ConvertedCompYearly'],
    errors='coerce'
)

In [92]:
top_earners_clean = top_earners.dropna()

In [94]:
top_5 = top_earners_clean.sort_values(
    by='ConvertedCompYearly',
    ascending=False
).head(5)

top_5

,EdLevel,Country,ConvertedCompYearly
34267,"Associate degree (A.A., A.S., etc.)",United States of America,50000000.0
28700,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Ukraine,33552715.0
43143,"Associate degree (A.A., A.S., etc.)",Poland,18387548.0
35353,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Iran, Islamic Republic of...",15430267.0
45971,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Netherlands,13921760.0


In [95]:
top_earners_filtered = top_earners_clean[
    top_earners_clean['ConvertedCompYearly'] < 500000
]

In [96]:
top_5 = top_earners_filtered.sort_values(
    by='ConvertedCompYearly',
    ascending=False
).head(5)

top_5

,EdLevel,Country,ConvertedCompYearly
12943,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Republic of Korea,499750.0
20932,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",United States of America,480000.0
22799,Some college/university study without earning ...,Romania,480000.0
11416,"Associate degree (A.A., A.S., etc.)",United States of America,480000.0
1787,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",United States of America,480000.0


### Insight

After removing extreme outliers, the highest-paid respondents have annual salaries close to 480,000–499,750 USD.

Among the top earners:
- Most hold a **Bachelor’s or Master’s degree**, indicating that higher education is common but not strictly required.
- One respondent has **some college education without completing a degree**, suggesting that formal education is not the only path to high income.
- The majority are located in the **United States**, with others in countries like **South Korea and Romania**.

This suggests that while higher education is common among top earners, **practical skills, experience, and location (especially working in high-paying markets like the US)** play a critical role in achieving high compensation.

**________________________________________________________________________________________________________________________________**

## 🎯 Bonus Question 1: What percentage of Python users are there in each age group?

### Approach

To calculate the percentage of Python users in each age group:

- First, we group all respondents by age.
- Then, we identify respondents who reported using Python.
- For each age group, we calculate:
  - Total number of respondents
  - Number of Python users
- Finally, we compute the percentage of Python users within each age group.

This helps us understand how Python usage varies across different age groups.

In [99]:
# Python users
python_users = df[df['LanguageHaveWorkedWith'].str.contains('Python', na=False)]

# Total respondents by age
age_total = df['Age'].value_counts()

# Python users by age
age_python = python_users['Age'].value_counts()

# Combine into one DataFrame
age_analysis = pd.DataFrame({
    'total': age_total,
    'python_users': age_python
}).fillna(0)

# Calculate percentage
age_analysis['python_percentage'] = (age_analysis['python_users'] / age_analysis['total']) * 100

age_analysis['python_percentage'] = age_analysis['python_percentage'].round(2)
# Sort
age_analysis_sorted = age_analysis.sort_values(by='python_percentage', ascending=False)

age_analysis_sorted.head(10)

,total,python_users,python_percentage
Age,,,
18-24 years old,9210,3684,40.00
45-54 years old,6275,2424,38.63
55-64 years old,2626,978,37.24
25-34 years old,16519,6102,36.94
35-44 years old,13241,4862,36.72
65 years or older,942,298,31.63
Prefer not to say,378,118,31.22


### Insight

Python usage is relatively consistent across most age groups, ranging between approximately 36% and 40%.

The highest adoption rate is observed among the 18–24 age group (~40%), indicating strong early adoption among younger developers.

Interestingly, older age groups (45–64) also show high Python usage (~37–39%), suggesting that Python is widely adopted across experience levels, not only by younger professionals.

The lowest adoption rate appears in the 65+ group (~31%), which may reflect lower engagement with newer technologies.

Overall, this indicates that Python is not limited to a specific age group but is broadly used across the developer population.

*________________________________________________________________________________________________________________________________*

## 🎯 Bonus Question 2: What industries are most common among remote workers in the top 25% salary group?

### Approach

To identify the most common industries among high-earning remote workers:

- First, we convert the salary column to numeric format and remove missing values.
- Then, we calculate the 75th percentile (top 25% salary threshold).
- We filter respondents whose salary is above this threshold.
- Next, we identify remote workers by filtering the RemoteWork column (including Remote and Hybrid).
- Finally, we analyze the Industry column to find the most common industries among this group.

This helps us understand which industries dominate among high-income remote professionals.

In [100]:
# Step 1: Prepare salary column
df['ConvertedCompYearly'] = pd.to_numeric(df['ConvertedCompYearly'], errors='coerce')

In [101]:
# Step 2: Remove missing salaries
salary_clean = df.dropna(subset=['ConvertedCompYearly'])

In [102]:
# Step 3: Calculate top 25% threshold
salary_threshold = salary_clean['ConvertedCompYearly'].quantile(0.75)

In [103]:
salary_threshold

np.float64(120596.0)

In [104]:
# Step 4: Filter high earners (top 25%)
high_earners = salary_clean[salary_clean['ConvertedCompYearly'] >= salary_threshold]

In [105]:
# Step 5: Filter remote workers (Remote + Hybrid)
remote_workers = high_earners[
    high_earners['RemoteWork'].str.contains('Remote|Hybrid', na=False)
]

In [106]:
# Step 6: Analyze industries
industry_counts = remote_workers['Industry'].value_counts()

industry_counts.head(10)

Industry
Software Development                          1820
Other:                                         343
Fintech                                        331
Healthcare                                     284
Internet, Telecomm or Information Services     247
Banking/Financial Services                     216
Government                                     150
Retail and Consumer Services                   135
Media & Advertising Services                   130
Transportation, or Supply Chain                117
Name: count, dtype: int64

In [108]:
# Industry percentage distribution
industry_percentage = (industry_counts / industry_counts.sum()) * 100
industry_percentage.round(2).head(10)

Industry
Software Development                          42.89
Other:                                         8.08
Fintech                                        7.80
Healthcare                                     6.69
Internet, Telecomm or Information Services     5.82
Banking/Financial Services                     5.09
Government                                     3.54
Retail and Consumer Services                   3.18
Media & Advertising Services                   3.06
Transportation, or Supply Chain                2.76
Name: count, dtype: float64

## Insight
Software Development clearly dominates both in count and percentage (~43%), making it the leading industry among high-income remote workers.

Other industries such as Fintech, Healthcare, and Telecommunications follow but with significantly lower shares, indicating a strong concentration of high-paying remote roles in tech-driven sectors.

This suggests that remote high-income opportunities are heavily clustered in software and technology-related fields, where digital infrastructure and global collaboration are essential.

*________________________________________________________________________________________________________________________________*